src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs

```diff
@@ -457,11 +461,22 @@ public sealed override IEnumerable<CombinedDependencyListEntry> GetConditionalSt
                         if (!isStaticInterfaceMethod && !needsDependenciesForInstanceInterfaceMethodImpls)
                             continue;
 
+                        MethodDesc interfaceMethodDefinition = interfaceMethod;
+                        if (interfaceType != interfaceDefinitionType)
+                            interfaceMethodDefinition = factory.TypeSystemContext.GetMethodForInstantiatedType(interfaceMethodDefinition.GetTypicalMethodDefinition(), (InstantiatedType)interfaceDefinitionType);
+
                         MethodDesc implMethod = isStaticInterfaceMethod ?
-                            defType.ResolveInterfaceMethodToStaticVirtualMethodOnType(interfaceMethod) :
-                            defType.ResolveInterfaceMethodToVirtualMethodOnType(interfaceMethod);
+                            defTypeDefinition.ResolveInterfaceMethodToStaticVirtualMethodOnType(interfaceMethodDefinition) :
+                            defTypeDefinition.ResolveInterfaceMethodToVirtualMethodOnType(interfaceMethodDefinition);
                         if (implMethod != null)
                         {
+                            TypeDesc implType = defType;
+                            while (!implType.HasSameTypeDefinition(implMethod.OwningType))
```

Comment:
So in theory we shouldn't need to. We'd just call `implMethod.InstantiateSignature(defType.Instantiation, Instantiation.Empty)`.

This however doesn't work due to a quirk. I've been thinking about doing something about it but also I don't really want to perturb things too much for the sake of making things prettier.

The problem is this:

The type system has a representation of a generic type definition and "an open type instantiation". One is `Foo<T, U>`. The other is `Foo<!0, !1>`. We need to be able to distinguish between these because of stuff like `class Foo<T, U> { void Do(T m) { typeof(Foo<>); typeof(Foo<T, U>); }` - these typeofs need to produce a different thing. The instantiation in the first case is over generic parameters (`GenericParameterDesc`) in the former case, and signature variables in the latter case

Now here's the quirk. When iterating methods on these types, we get `Do(T m)` and `Do(!0 m)` respectively. Calling `InstantiateSignature` on `Do(T m)` doesn't do anything. When one thinks about it, it really better shouldn't do anything on generic parameters because `InstantiateSignature` is also used to make things like specialized method bodies and that would coalesce the `typeof(Foo<>)` and `typeof(Foo<T, U>)` cases into the same thing.

Since we do the resolution on `Foo<T, U>`, we'll typically get `Foo<T, U>.Method` back. But sometimes also `Base<!0>.Method`. `InstantiateSignature` would work on the latter, but not on the former. So we have a bunch of code written like this that can handle both cases correctly.

Reply:
> class Foo<T, U> { void Do(T m) { typeof(Foo<>); typeof(Foo<T, U>); }

FWIW Roslyn has a similar distinction, but maybe it's subtly different. From Roslyn's perspective, `Foo<>` is a completely unbound generic, similar to a definition. I believe you'll get the same symbol from that as you would calling `OriginalDefinition` on any given instantiation of `Foo<T, U>`.

`typeof(Foo<T, U>)`, on the other hand, is in no way open or unbound in that instance. It's fully instantiated using the type variables `T` and `U`. In other words, when looking at the definition of the class `class Foo<T, U>`, if you look at `T` and `U` you will see that they are definition symbols -- they define `T` and `U`. If you look at `T` and `U` in `typeof(Foo<T, U>)`, however, you'll see that they are _type arguments_ -- they are references to the definitions and would be no different than if you said `typeof(Foo<int, string>)`.

I'm not sure if that helps anything. I think the Roslyn equivalent of something similar to the above would be calling `Construct` on an already constructed type symbol. Roslyn produces an exception if you try to do this. So maybe that's the same as NAOT. Roslyn would demand that you pass in the original definition before calling `Construct`.